ckanext/dcat/utils.py

```diff
@@ -0,0 +1,146 @@
+import logging
+
+from pylons import config
+
+from ckan import model
+
+log = logging.getLogger(__name__)
+
+
+def catalog_uri():
+    '''
+    Returns an URI for the whole catalog
+
+    This will be used to uniquely reference the CKAN instance on the RDF
+    serializations and as a basis for eg datasets URIs (if not present on
+    the metadata).
+
+    The value will be the first found of:
+
+        1. The `ckanext.dcat.base_uri` config option (recommended)
+        2. The `ckan.site_url` config option
+        3. `http://` + the `app_instance_uuid` config option (minus brackets)
+
+    A warning is emited if the third option is used.
+
+    Returns a string with the catalog URI.
+    '''
+
+    uri = config.get('ckanext.dcat.base_uri')
+    if not uri:
+        uri = config.get('ckan.site_url')
+    if not uri:
+        uri = 'http://' + config.get('app_instance_uuid').replace('{', '') \
+                                                         .replace('}', '')
+        log.critical('Using app id as catalog URI, you should set the ' +
+                     '`ckanext.dcat.base_uri` or `ckan.site_url` option')
+    return uri
+
+
+def dataset_uri(dataset_dict):
+    '''
+    Returns an URI for the dataset
+
+    This will be used to uniquely reference the dataset on the RDF
+    serializations.
+
+    The value will be the first found of:
+
+        1. The value of the `uri` field
+        2. The value of an extra with key `uri`
+        3. `catalog_uri()` + '/dataset/' + `id` field
```

Comment:
I would have thought package.name was preferable to package.id. I realise that the name might change and "cool URIs don't change". But it's pretty rare. And Linked Data people really like readable URIs, and my feeling is that that is more important. I'd love to get an 'expert' opinion on this though.